README.md

```diff
@@ -6,26 +6,81 @@ Implements https://groups.google.com/forum/#!msg/kubernetes-sig-api-machinery/rH
+## Running controller
 
-Prerequisite: Since the sample-controller uses apps/v1 deployments, the Kubernetes cluster version should be greater than 1.9.
+There're two ways to run `EtcdProxyController`:
 
+* in-cluster, which is done by deploying [the deployment manifest, located in the `artifacts/deployment` directory](https://github.com/xmudrii/etcdproxy-controller/blob/master/artifacts/deployment/00-etcdproxy-controller.yaml),
+* out-of-cluster, which is done by building the controller binary and running it on the local machine.
+
+Before running the controller, you need to deploy the etcd cluster. There's an example manifests along with the deploying instructions in the [`artifacts/etcd` directory](https://github.com/xmudrii/etcdproxy-controller/tree/master/artifacts/etcd).
+
+The deployment manifest assumes etcd is available on `https://etcd-svc-1.etcd.svc:2379`. This can be configured by updating the [`00-etcdproxy-controller.yaml` manifest](https://github.com/xmudrii/etcdproxy-controller/blob/master/artifacts/deployment/00-etcdproxy-controller.yaml).
+
+When running out-of-cluster, the etcd URL can be changed using the `etcd-core-url` flag.
+
+### Providing the core etcd client certificates to the controller
+
+When deploying the core etcd using the provided manifests, you can deploy the client certificates using the `etcd-client-certs.yaml` manifest. [The README of `artifacts/etcd`](https://github.com/xmudrii/etcdproxy-controller/tree/master/artifacts/etcd) contains more details about deploying the client certificates.
+
+If you're deploying etcd using another method, the `EtcdProxyController` requires you to provide it the CA certificate (`ca.pem`) as a ConfigMap, and the client certificate (`client.pem`) and key (`client-key.pem`) as
+a Secret.
+
+Currently, both the ConfigMap and Secret must be located in the controller namespace.
+
+The ConfigMap is by default called `etcd-coreserving-ca`, and the name can be changed using the `--etcd-core-ca-configmap` flag.
+The Secret is by default called `etcd-coreserving-cert`, and the name can be changed using the `--etcd-core-ca-secret` flag.
+
+The ConfigMap and Secret can be created using the following `kubectl` commands:
+```
+kubectl create configmap etcd-coreserving-ca --from-file=ca.pem -n kube-apiserver-storage
+kubectl create secret generic etcd-coreserving-cert --from-file=client.pem --from-file=client-key.pem -n kube-apiserver-storage
 ```
-# assumes you have a working kubeconfig, not required if operating in-cluster
-$ go run *.go -kubeconfig=$HOME/.kube/config
 
-# create a CustomResourceDefinition
-$ kubectl create -f artifacts/examples/etcdstorage-crd.yaml
+### Running out-of-cluster
 
-# create a custom resource of type Foo
-$ kubectl create -f artifacts/examples/etcdstorage-cr.yaml
+Before running the controller out-of-cluster you need to create the etcd proxy namespace, which is by default called `kube-apiserver-storage`, and can be configured using the `--namespace` flag:
```

Comment:
better: by default called `...` (can be customized using `--namespace`):

Reply:
Done.